// angularjs-2-localstorage-timer/src/app/app.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var broadcaster_service_1 = require("./broadcast/broadcaster.service");
var localstorage_service_1 = require("./localstorage/localstorage.service");
var timer_service_1 = require("./timer/timer.service");
var AppComponent = (function () {
    function AppComponent(broadcaster, AppTimer) {
        this.broadcaster = broadcaster;
        this.AppTimer = AppTimer;
        this.CLASS = "[AppComponent]";
        this.properties = {
            counter: 5,
            interval: 1000,
            id: "CountDownTest"
        };
        // Events              
        this.broadcaster.on('CountDownTimerChangeEvent').subscribe(function (message) {
            console.log("[event] CountDownTimerChangeEvent");
        });
        this.broadcaster.on('CountDownTimerEndEvent').subscribe(function (message) {
            console.log("[event] CountDownTimerEndEvent");
        });
        this.reload();
    }
    // Timer functionality
    AppComponent.prototype.reload = function () {
        this.AppTimer.init(this.properties);
    };
    AppComponent.prototype.recreate = function () {
        console.log(this.CLASS, "recreate with", this.properties);
        this.AppTimer.init(this.properties);
    };
    AppComponent.prototype.resume = function () {
        this.AppTimer.start();
    };
    AppComponent.prototype.destory = function () {
        this.AppTimer.destory();
    };
    return AppComponent;
}());
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], AppComponent.prototype, "properties", void 0);
AppComponent = __decorate([
    core_1.Component({
        selector: 'my-app',
        templateUrl: './app.component.html',
        providers: [broadcaster_service_1.Broadcaster, timer_service_1.CountDownTimer, localstorage_service_1.LocalStorageService]
    }),
    __metadata("design:paramtypes", [broadcaster_service_1.Broadcaster, timer_service_1.CountDownTimer])
], AppComponent);
exports.AppComponent = AppComponent;
