// angularjs-2-localstorage-timer/src/app/localstorage/localstorage.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var LocalStorageService = (function () {
    function LocalStorageService() {
        this.SERVICE = "[LocalStorage]";
        this.DEBUG = false; // Throw console msgs ? true/false
        this.id = "MyStorage";
        if (this.DEBUG) {
            console.log(this.SERVICE, "[init]");
        }
    }
    LocalStorageService.prototype.get = function () {
        // Get data        
        var data = JSON.parse(window.localStorage && window.localStorage.getItem(this.id));
        if (this.DEBUG) {
            console.log(this.SERVICE, "[get]", this.id, "=", data);
        }
        return data;
    };
    LocalStorageService.prototype.setId = function (id) {
        this.id = id;
    };
    LocalStorageService.prototype.set = function (data) {
        if (this.DEBUG) {
            console.log(this.SERVICE, "[set]", this.id, "=", data);
        }
        window.localStorage && window.localStorage.setItem(this.id, JSON.stringify(data));
    };
    LocalStorageService.prototype.destory = function () {
        window.localStorage && window.localStorage.removeItem(this.id);
    };
    return LocalStorageService;
}());
LocalStorageService = __decorate([
    core_1.Injectable(),
    __metadata("design:paramtypes", [])
], LocalStorageService);
exports.LocalStorageService = LocalStorageService;
